File: 98-validate-binary-search-tree/validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    if (!root) return true
    return validateBST(root, null, null)
};

function validateBST (root, min, max) {
    if (root === null) return true

    if ((min !== null && root.val <= min) || (max !== null && root.val >= max)) return false

    return validateBST(root.left, min, root.val) && validateBST(root.right, root.val, max)
}

File: 98-validate-binary-search-tree/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def is_valid_bst_helper(node, min_val, max_val):
            if not node:
                return True

            if not (min_val < node.val < max_val):
                return False

            left_is_valid = is_valid_bst_helper(node.left, min_val, node.val)

            right_is_valid = is_valid_bst_helper(node.right, node.val, max_val)

            return left_is_valid and right_is_valid

        return is_valid_bst_helper(root, -math.inf, math.inf)